### ISSUE TITLE
Is SSLFactory thread safe?

### ISSUE THREAD

**Issue by yingzhuo** [Hakky54/sslcontext-kickstart#534]
Hello, I like this lib very very much. It saved my hours, thank you.
I'm using v8.3.6 currently.

My question is simple: Is the instance of `nl.altindag.ssl.SSLFactory` thread safe or not?

Here is my code:

```java
@Component("restClientForBill")
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
public class ClientHttpRequestFactoryFactoryBean implements FactoryBean<RestClient> {

    // "bill" is Subject DN CommonName (CN)

    private static final String ID_CERT_PEM = "META-INF/client/client-bill-cert-x509.pem";
    private static final String ID_PRIVATE_PEM = "META-INF/client/client-bill-privatekey-pkcs8.pem";
    private static final char[] ID_PRIVATE_PEM_PASS = "123456".toCharArray();

    private static final String TRUST_CERT_PEM = "META-INF/client/trust-cert-x509.pem";

    private final Environment environment;

    public ClientHttpRequestFactoryFactoryBean(Environment environment) {
        this.environment = environment;
    }

    @Override
    public RestClient getObject() {
        var sslFactory = SSLFactory.builder()
                .withIdentityMaterial(
                        PemUtils.loadIdentityMaterial(ID_CERT_PEM, ID_PRIVATE_PEM, ID_PRIVATE_PEM_PASS)
                )
                .withTrustMaterial(
                        PemUtils.loadTrustMaterial(TRUST_CERT_PEM)
                )
                .build();

        var socketFactory = Apache5SslUtils.toSocketFactory(sslFactory);
        var connectionManager = PoolingHttpClientConnectionManagerBuilder.create()
                .setSSLSocketFactory(socketFactory)
                .build();

        var httpClient = HttpClients.custom()
                .setConnectionManager(connectionManager)
                .build();

        var requestFactory = new HttpComponentsClientHttpRequestFactory(httpClient);

        return RestClient.builder()
                .requestFactory(requestFactory)
                .baseUrl(environment.resolvePlaceholders("https://127.0.0.1:${server.port}/"))
                .defaultHeaders(headers -> {
                    headers.setAccept(List.of(MediaType.APPLICATION_JSON));
                    headers.setContentType(MediaType.APPLICATION_JSON);
                })
                .build();
    }

    @Override
    public Class<?> getObjectType() {
        return RestClient.class;
    }

}
```

I noticed that the [example code](https://github.com/Hakky54/mutual-tls-ssl/blob/35cba2f3a2dcd73b01fa323b99eec7777f7429bb/client/src/main/java/nl/altindag/client/SSLConfig.java#L27) is using `@Scope("prototype")`.

May I ask why?

**Comment by Hakky54** [comment]
> Hello, I like this lib very very much. It saved my hours, thank you.

Thank you for your kind words ❤️ 

> My question is simple: Is the instance of nl.altindag.ssl.SSLFactory thread safe or not?

I think so, but to be honest I did not do a proper testing.  I put the effort to make it thread safe such as using immutables, returning copies of internals when a get method is called and it has read/write locks in place when something can be changed in a concurrent fashion etc. Next to that SSLFactory acs as a wrapper for some default SSL classes of the java base module such as SSLContext, SSLEngine, SSLSocket etc which are thread safe. So it should be threadsafe, but I cannot confirm it with 100% as I didn't tested that usecase.

> I noticed that the [example code](https://github.com/Hakky54/mutual-tls-ssl/blob/35cba2f3a2dcd73b01fa323b99eec7777f7429bb/client/src/main/java/nl/altindag/client/SSLConfig.java#L27) is using @Scope("prototype").

I added that 4 years ago when the SSLFactory existed in a different form within that project, see here: https://github.com/Hakky54/mutual-tls-ssl/commit/ae75f5fdb5b6b3851538eef282988976145a85f8 It was more in a beta form. Back then it was somehow failing to use with multiple http clients and therefore I added prototype scope. Later I created this library out of that basic ssl configuration. The library evolved and got more mature including thread safe operations and other functionalities such as hot reloading. I never thought of removing the prototype scope annotation. I just did now in a separate pull request here: https://github.com/Hakky54/mutual-tls-ssl/pull/107 and it seems that it is also working without that.

I hope this give you a better understanding why the prototype scope was added initially and that it is not needed anymore.
